web/controllers/api/release_controller.ex

```diff
@@ -118,23 +117,20 @@ defmodule HexWeb.API.ReleaseController do
     end
   end
 
-  defp create(conn, package, release_params, checksum, user, body) do
-    result =
-      HexWeb.Repo.transaction(fn ->
-        case Release.create(package, release_params, checksum) do
-          {:ok, release} ->
-            audit(conn, "release.publish", {package, release})
-            release
-          {:error, changeset} ->
-            HexWeb.Repo.rollback(changeset)
-        end
-      end)
+  defp create(conn, package, params, checksum, user, body) do
+    params = normalize_params(params)
 
-    case result do
-      {:ok, release} ->
+    multi =
+      Ecto.Multi.new
+      |> Ecto.Multi.insert(:release, Release.create(package, params, checksum))
+      |> Ecto.Multi.run(:log, fn %{release: release} -> audit(user, "release.publish", {package, release}) end)
+
+    case HexWeb.Repo.transaction(multi) do
+      {:ok, %{release: release}} ->
         after_release(package, release.version, user, body)
 
         release = %{release | package: package}
+        release = HexWeb.Repo.preload(release, :requirements)
```

Comment:
I needed to preload requirements because that association was not loaded after creation when requirements were empty. In other cases it was loaded. See: https://github.com/elixir-lang/ecto/issues/1305